client-js/app/elements/labrad-registry.ts

```diff
@@ -129,6 +129,101 @@ export class LabradRegistry extends polymer.Base {
     );
   }
 
+  //Drag and drop
+  @listen('dragenter')
+  overBehaviour(event) {
+    event.preventDefault(); //I don't understand this.
+  }
+
+  @listen('dragover')
+  changeBehaviour(event) {
+    //allows the ctrl key to be pressed to change cursor between copy/move
+    event.preventDefault(); //I don't understand this. But needs to be here
+
+    if (event.ctrlKey) {
+      event.dataTransfer.dropEffect = "copy";
+    }
+    else {
+      event.dataTransfer.dropEffect = "move";
+    }
+  }
+
+  @listen('dragleave')
+  leaveDrag(event) {
+    event.preventDefault(); //I don't understand this. But it needs to be here
+  }
+
+  @listen('dragstart')
+  startDrag(event) {
+    //detect start of drag event, grab info about target
+    var data: any; //I tried enum, but kept getting errors...
+    data = {path: this.path, name: event.target.name};
+    event.dataTransfer.setData('text', JSON.stringify(data));
+    event.dataTransfer.effectAllowed = 'copyMove';
+    this.$.toastText.text = "drag to move, hold ctrl to copy";
+    this.$.toastText.show();
+  }
+
+  @listen('dragend')
+  endDrag(event) {
+    console.log('drag ended',event);
+  }
+
+  dirDrop(event) {
+    //behaviour for dropping on folders
+    console.log(event);
+    var data: any;
+    data = JSON.parse(event.dataTransfer.getData('text'));
```

Comment:
You can just say
```
var data = JSON.parse(event.dataTransfer.getData('text'));
```
The compiler will infer a type of any (which I presume is what `JSON.parse` returns).